64_FFT/main.cpp

```diff
@@ -0,0 +1,352 @@
+// Copyright (C) 2018-2024 - DevSH Graphics Programming Sp. z O.O.
+// This file is part of the "Nabla Engine".
+// For conditions of distribution and use, see copyright notice in nabla.h
+
+
+// I've moved out a tiny part of this example into a shared header for reuse, please open and read it.
+#include "nbl/application_templates/MonoDeviceApplication.hpp"
+#include "nbl/application_templates/MonoAssetManagerAndBuiltinResourceApplication.hpp"
+
+
+using namespace nbl;
+using namespace core;
+using namespace system;
+using namespace asset;
+using namespace video;
+
+
+#include "app_resources/common.hlsl"
+#include "nbl/builtin/hlsl/bit.hlsl"
+#include "nbl/builtin/hlsl/random/xoroshiro.hlsl"
+
+
+// Simple showcase of how to run FFT on a 1D array
+class FFT_Test final : public application_templates::MonoDeviceApplication, public application_templates::MonoAssetManagerAndBuiltinResourceApplication
+{
+	using device_base_t = application_templates::MonoDeviceApplication;
+	using asset_base_t = application_templates::MonoAssetManagerAndBuiltinResourceApplication;
+
+	smart_refctd_ptr<IGPUComputePipeline> m_pipeline;
+
+	smart_refctd_ptr<nbl::video::IUtilities> m_utils;
+
+	nbl::video::StreamingTransientDataBufferMT<>* m_upStreamingBuffer;
+	StreamingTransientDataBufferMT<>* m_downStreamingBuffer;
+	smart_refctd_ptr<nbl::video::IGPUBuffer> m_deviceLocalBuffer;
+
+	// These are Buffer Device Addresses
+	uint64_t m_upStreamingBufferAddress;
+	uint64_t m_downStreamingBufferAddress;
+	uint64_t m_deviceLocalBufferAddress;
+
+	// You can ask the `nbl::core::GeneralpurposeAddressAllocator` used internally by the Streaming Buffers give out offsets aligned to a certain multiple (not only Power of Two!)
+	uint32_t m_alignment;
+
+	// This example really lets the advantages of a timeline semaphore shine through!
+	smart_refctd_ptr<ISemaphore> m_timeline;
+	uint64_t semaphorValue = 0;
+
+	inline core::smart_refctd_ptr<video::IGPUShader> createShader(
+		const char* includeMainName)
+	{
+
+		const char* sourceFmt =
+			R"===(
+		#include "%s"
+		)===";
+
+		const size_t extraSize = 4u + 4u + 26u + 128u;
+
+		auto shader = core::make_smart_refctd_ptr<ICPUBuffer>(strlen(sourceFmt) + extraSize + 1u);
+		snprintf(
+			reinterpret_cast<char*>(shader->getPointer()), shader->getSize(), sourceFmt,
+			includeMainName
+		);
+
+		auto CPUShader = core::make_smart_refctd_ptr<ICPUShader>(std::move(shader), IShader::E_SHADER_STAGE::ESS_COMPUTE, IShader::E_CONTENT_TYPE::ECT_HLSL, includeMainName);
+		assert(CPUShader);
+		return m_device->createShader(CPUShader.get());
```

Comment:
clean that up into something saner, a-la https://github.com/Devsh-Graphics-Programming/Nabla/blob/bindless_blit/src/nbl/video/utilities/CComputeBlit.cpp#L68